### ISSUE TITLE
Mimic the SpeechSynthesis API on the simulation window

### ISSUE THREAD

**Issue by jessegreenberg** [phetsims/fenster#3]
From https://github.com/phetsims/utterance-queue/issues/63 - To forward requests from the simulation to the Cordova app we need to implement and respond to speak from the simulation. We also need the SpeechSynthesis interface to be available in the simulation as it loads so Voicing features are initialized on startup.

It seemed that we couldn't set objects to the simulation window but @zepumph showed a good way that works! In general, a saved variable reference to the simulation contentWindow object cannot be trusted, go through `simFrame.contentWindow` getter instead. The `simFrame.onload` event is a good time to set the SpeechSynthesis object because it is when the simFrame is ready and before the simulation loads.

Lets try an implementation of SpeechSynthesis!

**Comment by jessegreenberg** [comment]
I am finding that the `simFrame.onload` event is still too late to set the SpeechSynthesis API for usage in the sim as it initializes. I tried to use this: https://stackoverflow.com/questions/13270290/how-to-execute-code-before-window-load-and-after-dom-has-been-loaded and the Document.readyState to detect when the sim frame document is ready but the window has noet yet begun to load, but I am not seeing the `onreadystatechange` event fire in Cordova.

EDIT: According to https://stackoverflow.com/questions/19384661/iframe-onreadystatechange-function, the onreadystatechange is bad. I tried the `DOMContentLoaded` on the document event instead on the `simFrame.contentWindow.document` and still didn't see it fire.

EDIT: I tried this, setting simFrame.contentWindow.SpeechSynthesis outside of a load event. In this case `simFrame.contentWindow.speechSynthesis` is not defined in the sim either on sim initialization OR after the sim load event. When using the `load` event of the sim frame at least simFrame.contentWindow.speechSynthesis is defined by the time the sim load event is posted.

<details>

```js
  simFrame.contentWindow.SpeechSynthesis = SpeechSynthesis;
  simFrame.contentWindow.SpeechSynthesisUtterance = SpeechSnthesisUtterance;

  // the global instance to speak
  simFrame.contentWindow.speechSynthesis = new SpeechSynthesis();

  console.log( 'window objects set' );
  simFrame.contentWindow.speechSynthesis.speak( new simFrame.contentWindow.SpeechSynthesisUtterance( 'Speaking through window API' ) );

  simFrame.src = 'quadrilateral_en_phet.html?postMessageOnLoad';
```
</details>

EDIT: I tried the same code as above, but setting the `src` before the `simFrame.contentWindow.*` objects. 

EDIT: I found that the `reportColor` event happens before the PreferencesConfiguration is used to enable Voicing. So I tried to  set up objects on the window the first time that we receive that event like this:
```js
    // HACK ALERT!! We get the reportColor before we check the preferencesConfiguration for the existence of
    // window.speechSynthesis to enable voicing. So when we get this event, the simFrame.contentWindow should be
    // "stable" and we can set globals to it for usage in the sim.
    if ( data.type === 'reportColor' && simFrame.contentWindow.speechSynthesis === undefined ) {
      simFrame.contentWindow.SpeechSynthesis = SpeechSynthesis;
      simFrame.contentWindow.SpeechSynthesisUtterance = SpeechSnthesisUtterance;

      // the global instance to speak
      simFrame.contentWindow.speechSynthesis = new SpeechSynthesis();

      console.log( simFrame.contentWindow.phet.utteranceQueue.SpeechSynthesisAnnouncer.isSpeechSynthesisSupported() );
      simFrame.contentWindow.speechSynthesis.speak( new simFrame.contentWindow.SpeechSynthesisUtterance( 'Speaking through window API' ) );
    }
```

The behavior I am seeing is very odd...
<details>

```
[INFO:CONSOLE(39)] "message received reportColor", source: https://localhost/js/index.js (39)
[INFO:CONSOLE(51)] "true", source: https://localhost/js/index.js (51)
[INFO:CONSOLE(131)] "Speaking through window API", source: https://localhost/js/index.js (131)
[INFO:CONSOLE(39)] "message received reportColor", source: https://localhost/js/index.js (39)
[INFO:CONSOLE(39)] "message received reportColor", source: https://localhost/js/index.js (39)
[INFO:CONSOLE(39)] "message received supportsVoicing: true, isSpeechSynthesisSupported: false, locale: true", source: https://localhost/js/index.js (39)
[INFO:CONSOLE(39)] "message received speechSynthesis: undefined, utterance: undefined", source: https://localhost/js/index.js (39)
[INFO:CONSOLE(39)] "message received load", source: https://localhost/js/index.js (39)
[INFO:CONSOLE(58)] "true", source: https://localhost/js/index.js (58)
```

First message we get the first `reportColor` postMessage from the sim. Second line, is checking `simFrame.contentWindow.phet.utteranceQueue.SpeechSynthesisAnnouncer.isSpeechSynthesisSupported()` and returns true! Third line is a call to 
```js
simFrame.contentWindow.speechSynthesis.speak( new simFrame.contentWindow.SpeechSynthesisUtterance( 'Speaking through window API' ) );
```
and it is successful. Next two are duplicates of `reportColor`. The next line shows that `SpeechSynthesisAnnouncer.isSpeechSynthesisSupported is false! And that `window.speechSynthesis` and `window.speechSynthesisUtterance` is undefined. Finally, on line 58 we confirm that 
```
console.log( simFrame.contentWindow.phet.utteranceQueue.SpeechSynthesisAnnouncer.isSpeechSynthesisSupported() );
```
is indeed true...
</details>

EDIT: I put
```js
    window.parent && window.parent.postMessage( JSON.stringify( {
      type: `speechSynthesis: ${window.speechSynthesis}, utterance: ${window.SpeechSynthesisUtterance}`,
      url: window.location.href
    } ), '*' );
```

In the `SpeechSynthesisAnnouncer.isSpeechSynthesisSupprted` function and I see the same behavior, both undefined the first time they are needed in PreferencesConfiguration, but ready for use after that.

EDIT: I think this is behaving oddly because postMessage is not synchronous, so we cannot guarantee the order of things.

**Comment by jessegreenberg** [comment]
As a point of success, I was able to see
```
console.log( simFrame.contentWindow.phet.utteranceQueue.SpeechSynthesisAnnouncer.isSpeechSynthesisSupported() );
```

print true after the simulation had loaded. But I haven't gotten it to set yet before the sim has started loading. I really want to figure this out because if we can set it this way the API is fully contained/implemented in fenster and there is no need for `postMessage` calls at all from the simulation.

**Comment by jessegreenberg** [comment]
At this point I am giving up on setting something on the sim frame window before the simulation initializes. But I really like that https://github.com/phetsims/fenster/issues/3#issuecomment-1047301345 worked AFTER sim load and I want to leverage that. If we can still implement SpeechSynthesis in fenster there is no need for `postMessage` between sim and cordova because the SpeechSynthesis that we define in fenster is used directly by the sim.

I am going to try to implement an interface of SpeechSynthesis in utterance-queue, that gets set on window.speechSynthesis when requested (probably by query parameter). The existince of this and an interface will enable Voicing to be enabled and components to load without error. Then after load we can set the globals to the implementation we provide in fenster.

**Comment by zepumph** [comment]
If you have to have this in the sim frame, I would recommend writing a new preload called something like `polyfillSpeechSynthesisWithParentPostMessage`. 


A couple other ideas for keeping this in fenster:
1. I would recommend testing in a limited frame if you continue. Sometimes it is hard to separate incorrect DOM actions vs Cordova problems. 
2. (I don't think this is that helpful), but as a stepping stone to a solution, and understanding cordova better, perhaps it makes sense to try to put an intermediate iframe in between cordova and the sim, and see if you have inject speechsythesis into quad from that frame. 


I took a stab at a limited case, thinking that `DOMContentLoaded` is the best path forward, But I couldn't get that injection to work. Though I think it could work, I have to give up. `iframe.html` is just a basic `body` with an `h3`.

```html
<iframe id="hello_goose"></iframe>

<script>

  const x = document.getElementById( 'hello_goose' );

  x.contentWindow.hi = true;
  console.log( '1', x.contentWindow.hi );
  x.src = './iframe.html';


  console.log( '1', x.contentWindow.hi );
  x.contentWindow.addEventListener( 'DOMContentLoaded', () => {
    debugger; // not getting this
  } );
  x.contentDocument.addEventListener( 'DOMContentLoaded', () => {
    debugger;// not getting this
  } );

  setTimeout( () => {
    console.log( '2', x.contentWindow.hi );
  }, 0 );

</script>


Good luck!

**Comment by jessegreenberg** [comment]
OK thanks! Great ideas, Ill do a little exploring in a more basic example. That will remove time taken by `cordova build`. I was about to bail and got started adding code to utterance-queue that would be behind a query parameter. But I couldn't get UtteranceQueue to work as a preload there. Ill try a little more to get `DOMContentLoaded` working. I agree it seems like that would be the injection point if possible.

**Comment by jessegreenberg** [comment]
I was able to get the DOMContentLoaded event by creating an iframe element and appending it to the body like this
```js
  frameElement.contentWindow.addEventListener( 'DOMContentLoaded', () => {
    // frameElement.contentWindow.hi = true;
    debugger; // not getting this
  } );
  console.log( frameElement.contentWindow.hi );

  document.body.appendChild( frameElement );
  frameElement.src = './iframe.html';
```
But at the debugger the `contentWindow` object of the frameElement is null.

**Comment by zepumph** [comment]
This is weird, but can you try to set an ID on frameElement, and then regrab the reference to the element in the debugger, will getElementById. For some reason I feel like it could be weird because of how the appending occurs.

**Comment by zepumph** [comment]
https://stackoverflow.com/questions/12199797/why-is-iframe-contentwindow-null actually isn't surprised that the window isn't ready on DOMContentLoaded. 

Perhaps we need to think about polling or investigating more why/how phet code has loaded before the 'load' event fires.

**Comment by jessegreenberg** [comment]
I put in a polyfill in utterance-queue that is conditionally loaded with a query parameter (though not as a preload yet). But it didn't work. It seems that I cannot set a function with code from a parent window to be run in a child window, which I verified in a basic case:

iframe HTML:
```html
<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <title>Hello!</title>
</head>
<body>
<h3>Im an iframe!</h3>
</body>

<script>

  window.myTestFunction = () => {
    console.log( 'test function implementation in iframe' );
  }

  window.setInterval( () => {
    window.myTestFunction();
  } );
</script>
</html>
```

parent HTML:
```html
<iframe id="hello-goose"></iframe>

<script>

  const frameElement = document.getElementById( 'hello-goose' );
  frameElement.src = './iframe.html';

  const parentImplementation = () => {
    console.log( 'parent implementation' );
  }

  frameElement.contentWindow.myTestFunction = parentImplementation;

</script>
```

It always prints the "test function implementation in iframe". I could see this being important for security reasons. A way to call code from the parent window is to use `window.parent`. In this example it always prints "parent implementation"

iframe:
```html
<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <title>Hello!</title>
</head>
<body>
<h3>Im an iframe!</h3>
</body>

<script>

  window.myTestFunction = () => {
    console.log( 'test function implementation in iframe' );
  }

  window.setInterval( () => {
    window.parent.myTestFunction();
  } );
</script>
</html>
```

parent html:
```html
<iframe id="hello-goose"></iframe>

<script>

  const frameElement = document.getElementById( 'hello-goose' );
  frameElement.src = './iframe.html';

  const parentImplementation = () => {
    console.log( 'parent implementation' );
  }

  window.myTestFunction = parentImplementation;

</script>
```

**Comment by jessegreenberg** [comment]
We think all of this work with Fenster is needed because SpeechSynthesis is not available in an Android WebView because of this Chromium bug: https://bugs.chromium.org/p/chromium/issues/detail?id=487255. Just referencing it here for documentation.

**Comment by jessegreenberg** [comment]
@jbphet and I worked on this more today with consultation from @zepumph. It may be possible to set the implementation on the child frame from the parent frame, but the timing has to be right - it has to be set when the child window is ready and before the child window begins to load resources and make usages of SpeechSynthesis. We found a working solution where we we pull in an implementation of SpeechSynthesis in the child frame (sim) from the parent frame (cordova).

The above commits accomplish this. When requested with a query parameter a `SpeechSynthesisParentPolyfill` is used in the sim to assign a different implementation of SpeechSynthesis to the window. The implementation is in the parent frame and accessed through `window.parent`.

Enough basics of the API are in place to trigger successful speech in the Cordova app.  But we need to finish the implementation and make sure that all features of Voicing (and the utterance-queue) are supported.

**Comment by jessegreenberg** [comment]
I verified that all properties of `SpeechSynthesisUtterance` work from the simulation including things like rate and pitch. Only exception is `volume` which is not supported by the cordova TTS plugin. But it isn't modified by sims either.

Changing voices is not supported yet.

In the above commit I made it so that Utterances do not cancel eachother by default to match the web specification. SpeechSynthesisAnnouncer manually cancels things for us when necessary. I did some testing and saw voicing responses as I would expect.

The voice rate is very slow. It sounds best on my phone at 2x.

I am seeing an issue with async cancel/end events, worthy of its own issue.
